File: ex24/imageStitcher.py
import os
import argparse
import cv2
import numpy as np

def compute_edge_density(image):
    """Compute edge density using Canny edge detector."""
    gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    edges = cv2.Canny(gray, 100, 200)
    return np.sum(edges > 0) / (image.shape[0] * image.shape[1])

def compute_color_histogram(image):
    """Compute color histogram for an image."""
    hist = []
    for i in range(3):
        hist.extend(cv2.calcHist([image], [i], None, [8], [0, 256]).flatten())
    return hist

def validate_image(image, reference_histograms, reference_edge_densities, hist_threshold=0.5, edge_threshold=0.5):
    """
    Validate if an image belongs to the set based on color histogram and edge density.
    """
    hist = compute_color_histogram(image)
    edge_density = compute_edge_density(image)
    
    for ref_hist, ref_edge_density in zip(reference_histograms, reference_edge_densities):
        hist_similarity = cv2.compareHist(np.array(hist, dtype=np.float32), 
                                         np.array(ref_hist, dtype=np.float32), 
                                         cv2.HISTCMP_CORREL)
        edge_diff = abs(edge_density - ref_edge_density) / max(edge_density, ref_edge_density)
        edge_similarity = 1 - edge_diff
        
        if hist_similarity > hist_threshold and edge_similarity > edge_threshold:
            return True
            
    return False

def calculate_reference_features(images, sample_size=5):
    """
    Calculate reference features from a subset of images.
    """
    if len(images) <= sample_size:
        sample_indices = range(len(images))
    else:
        sample_indices = np.random.choice(len(images), sample_size, replace=False)
    
    reference_histograms = []
    reference_edge_densities = []
    
    for idx in sample_indices:
        reference_histograms.append(compute_color_histogram(images[idx]))
        reference_edge_densities.append(compute_edge_density(images[idx]))
    
    return reference_histograms, reference_edge_densities

def stitch_images(images):
    """
    Stitch multiple images into a panorama.
    """
    stitcher = cv2.Stitcher_create()
    status, result = stitcher.stitch(images)
    
    if status != cv2.Stitcher_OK:
        raise Exception(f"Stitching failed with error code: {status}")
        
    return result

def main():
    parser = argparse.ArgumentParser(description="Stitch multiple images into a panorama with validation")
    parser.add_argument("-i", "--input", required=True, help="Input directory containing images")
    parser.add_argument("-o", "--output", required=True, help="Output file for the stitched panorama")
    parser.add_argument("--hist-threshold", type=float, default=0.6, help="Threshold for histogram similarity (0-1)")
    parser.add_argument("--edge-threshold", type=float, default=0.5, help="Threshold for edge density similarity (0-1)")
    
    args = parser.parse_args()
    
    if not os.path.isdir(args.input):
        print(f"Error: Input directory '{args.input}' does not exist.")
        return
    
    image_files = [f for f in os.listdir(args.input) if f.lower().endswith(('.png', '.jpg', '.jpeg', '.bmp', '.tiff'))]
    
    if not image_files:
        print(f"Error: No images found in '{args.input}'.")
        return
    
    print(f"Found {len(image_files)} images.")
    
    images = []
    for image_file in image_files:
        image_path = os.path.join(args.input, image_file)
        img = cv2.imread(image_path)
        if img is not None:
            images.append(img)
        else:
            print(f"Warning: Could not load image '{image_file}'.")
    
    if len(images) < 2:
        print("Error: Need at least 2 valid images for stitching.")
        return
    
    print("Calculating reference features...")
    reference_histograms, reference_edge_densities = calculate_reference_features(images)
    
    valid_images = []
    for i, img in enumerate(images):
        if i < len(reference_histograms):
            valid_images.append(img)
            continue
            
        if validate_image(img, reference_histograms, reference_edge_densities, 
                         args.hist_threshold, args.edge_threshold):
            valid_images.append(img)
            print(f"Image {i+1}/{len(images)}: Valid")
        else:
            print(f"Image {i+1}/{len(images)}: Invalid (filtered out)")
    
    print(f"{len(valid_images)} out of {len(images)} images passed validation.")
    
    if len(valid_images) < 2:
        print("Error: Not enough valid images for stitching after filtering.")
        return
    
    print("Stitching images...")
    try:
        result = stitch_images(valid_images)
        cv2.imwrite(args.output, result)
        print(f"Stitched image saved to '{args.output}'.")
    except Exception as e:
        print(f"Stitching failed: {e}")

if __name__ == "__main__":
    main()

File: ex24/imageStitcher.cpp
#include <iostream>
#include <vector>
#include <string>
#include <filesystem>
#include <random>
#include <algorithm>
#include <opencv2/opencv.hpp>
#include <opencv2/stitching.hpp>

namespace fs = std::filesystem;

// Compute edge density using Canny edge detector
double computeEdgeDensity(const cv::Mat& image) {
    cv::Mat gray, edges;
    cv::cvtColor(image, gray, cv::COLOR_BGR2GRAY);
    cv::Canny(gray, edges, 100, 200);
    
    int edgeCount = cv::countNonZero(edges);
    double totalPixels = static_cast<double>(image.rows * image.cols);
    return edgeCount / totalPixels;
}

// Compute color histogram for an image
std::vector<float> computeColorHistogram(const cv::Mat& image) {
    std::vector<float> hist;
    std::vector<cv::Mat> bgr_planes;
    cv::split(image, bgr_planes);
    
    int histSize = 8;
    float range[] = {0, 256};
    const float* histRange = {range};
    
    for (const auto& plane : bgr_planes) {
        cv::Mat hist_channel;
        cv::calcHist(&plane, 1, 0, cv::Mat(), hist_channel, 1, &histSize, &histRange);
        hist.insert(hist.end(), hist_channel.begin<float>(), hist_channel.end<float>());
    }
    
    return hist;
}

// Validate if an image belongs to the set based on color histogram and edge density
bool validateImage(const cv::Mat& image, 
                  const std::vector<std::vector<float>>& referenceHistograms,
                  const std::vector<double>& referenceEdgeDensities,
                  double histThreshold = 0.5,
                  double edgeThreshold = 0.5) {
    
    auto hist = computeColorHistogram(image);
    double edgeDensity = computeEdgeDensity(image);
    
    for (size_t i = 0; i < referenceHistograms.size(); i++) {
        double histSimilarity = cv::compareHist(cv::Mat(hist), cv::Mat(referenceHistograms[i]), cv::HISTCMP_CORREL);
        double edgeDiff = std::abs(edgeDensity - referenceEdgeDensities[i]) / std::max(edgeDensity, referenceEdgeDensities[i]);
        double edgeSimilarity = 1.0 - edgeDiff;
        
        if (histSimilarity > histThreshold && edgeSimilarity > edgeThreshold) {
            return true;
        }
    }
    
    return false;
}

// Calculate reference features from a subset of images
std::pair<std::vector<std::vector<float>>, std::vector<double>> 
calculateReferenceFeatures(const std::vector<cv::Mat>& images, int sampleSize = 5) {
    std::vector<std::vector<float>> referenceHistograms;
    std::vector<double> referenceEdgeDensities;
    
    std::vector<int> sampleIndices;
    if (images.size() <= static_cast<size_t>(sampleSize)) {
        for (size_t i = 0; i < images.size(); i++) {
            sampleIndices.push_back(i);
        }
    } else {
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_int_distribution<int> dist(0, images.size() - 1);
        
        while (sampleIndices.size() < static_cast<size_t>(sampleSize)) {
            int idx = dist(gen);
            if (std::find(sampleIndices.begin(), sampleIndices.end(), idx) == sampleIndices.end()) {
                sampleIndices.push_back(idx);
            }
        }
    }
    
    for (int idx : sampleIndices) {
        referenceHistograms.push_back(computeColorHistogram(images[idx]));
        referenceEdgeDensities.push_back(computeEdgeDensity(images[idx]));
    }
    
    return {referenceHistograms, referenceEdgeDensities};
}

// Stitch multiple images into a panorama
cv::Mat stitchImages(const std::vector<cv::Mat>& images) {
    cv::Ptr<cv::Stitcher> stitcher = cv::Stitcher::create();
    cv::Mat result;
    cv::Stitcher::Status status = stitcher->stitch(images, result);
    
    if (status != cv::Stitcher::OK) {
        throw std::runtime_error("Stitching failed with error code: " + std::to_string(status));
    }
    
    return result;
}

int main(int argc, char** argv) {
    cv::CommandLineParser parser(argc, argv,
        "{help h||show help message}"
        "{i|input|input directory containing images}"
        "{o|output|output file for the stitched panorama}"
        "{hist-threshold|0.6|threshold for histogram similarity (0-1)}"
        "{edge-threshold|0.5|threshold for edge density similarity (0-1)}"
    );
    
    if (parser.has("help") || !parser.has("i") || !parser.has("o")) {
        std::cout << "Usage: " << argv[0] << " -i=<input_dir> -o=<output_file> [--hist-threshold=<val>] [--edge-threshold=<val>]" << std::endl;
        parser.printMessage();
        return 0;
    }
    
    std::string inputDir = parser.get<std::string>("i");
    std::string outputFile = parser.get<std::string>("o");
    double histThreshold = parser.get<double>("hist-threshold");
    double edgeThreshold = parser.get<double>("edge-threshold");
    
    if (!fs::exists(inputDir) || !fs::is_directory(inputDir)) {
        std::cerr << "Error: Input directory '" << inputDir << "' does not exist." << std::endl;
        return 1;
    }
    
    std::vector<cv::Mat> images;
    std::vector<std::string> validExtensions = {".png", ".jpg", ".jpeg", ".bmp", ".tiff"};
    
    for (const auto& entry : fs::directory_iterator(inputDir)) {
        if (entry.is_regular_file()) {
            std::string ext = entry.path().extension().string();
            std::transform(ext.begin(), ext.end(), ext.begin(), ::tolower);
            
            if (std::find(validExtensions.begin(), validExtensions.end(), ext) != validExtensions.end()) {
                cv::Mat img = cv::imread(entry.path().string());
                if (!img.empty()) {
                    images.push_back(img);
                } else {
                    std::cerr << "Warning: Could not load image '" << entry.path().filename().string() << "'." << std::endl;
                }
            }
        }
    }
    
    if (images.size() < 2) {
        std::cerr << "Error: Need at least 2 valid images for stitching." << std::endl;
        return 1;
    }
    
    std::cout << "Found " << images.size() << " images." << std::endl;
    
    std::cout << "Calculating reference features..." << std::endl;
    auto [referenceHistograms, referenceEdgeDensities] = calculateReferenceFeatures(images);
    
    std::vector<cv::Mat> validImages;
    for (size_t i = 0; i < images.size(); i++) {
        if (i < referenceHistograms.size()) {
            validImages.push_back(images[i]);
            continue;
        }
        
        if (validateImage(images[i], referenceHistograms, referenceEdgeDensities, 
                         histThreshold, edgeThreshold)) {
            validImages.push_back(images[i]);
            std::cout << "Image " << i+1 << "/" << images.size() << ": Valid" << std::endl;
        } else {
            std::cout << "Image " << i+1 << "/" << images.size() << ": Invalid (filtered out)" << std::endl;
        }
    }
    
    std::cout << validImages.size() << " out of " << images.size() << " images passed validation." << std::endl;
    
    if (validImages.size() < 2) {
        std::cerr << "Error: Not enough valid images for stitching after filtering." << std::endl;
        return 1;
    }
    
    std::cout << "Stitching images..." << std::endl;
    try {
        cv::Mat result = stitchImages(validImages);
        cv::imwrite(outputFile, result);
        std::cout << "Stitched image saved to '" << outputFile << "'." << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "Stitching failed: " << e.what() << std::endl;
        return 1;
    }
    
    return 0;
}